did_peer/src/peer_did/peer_did/mod.rs

```diff
@@ -0,0 +1,223 @@
+pub mod generic;
+
+use core::fmt;
+use std::{fmt::Display, marker::PhantomData};
+
+use crate::error::DidPeerError;
+use did_parser::Did;
+use serde::{
+    de::{self, Visitor},
+    Deserialize, Deserializer, Serialize, Serializer,
+};
+
+use super::numalgos::{
+    numalgo3::Numalgo3,
+    traits::{Numalgo, ToNumalgo3},
+};
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct PeerDid<N: Numalgo> {
+    did: Did,
+    numalgo: N,
+}
+
+impl<N: Numalgo> PeerDid<N> {
+    pub fn parse<T>(did: T) -> Result<PeerDid<N>, DidPeerError>
+    where
+        Did: TryFrom<T>,
+        <Did as TryFrom<T>>::Error: Into<DidPeerError>,
+    {
+        N::parse(did)
+    }
+
+    pub fn did(&self) -> &Did {
+        &self.did
+    }
+
+    pub(crate) fn from_parts(did: Did, numalgo: N) -> PeerDid<N> {
+        Self { did, numalgo }
+    }
+}
+
+impl<N: ToNumalgo3> PeerDid<N> {
+    pub fn to_numalgo3(&self) -> Result<PeerDid<Numalgo3>, DidPeerError> {
+        self.numalgo.to_numalgo3(self.did())
+    }
+}
+
+impl<N: Numalgo> Serialize for PeerDid<N> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(self.did.did())
+    }
+}
+
+impl<'de, N: Numalgo> Deserialize<'de> for PeerDid<N> {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct PeerDidVisitor<N: Numalgo>(PhantomData<N>);
+
+        impl<'de, N: Numalgo> Visitor<'de> for PeerDidVisitor<N> {
+            type Value = PeerDid<N>;
+
+            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+                formatter.write_str("a string representing a DID")
+            }
+
+            fn visit_str<E>(self, value: &str) -> Result<Self::Value, E>
+            where
+                E: de::Error,
+            {
+                match N::parse(value.to_string()) {
+                    Ok(peer_did) => Ok(peer_did),
+                    Err(err) => Err(E::custom(format!("Failed to parse numalgo: {err}"))),
+                }
+            }
+        }
+
+        deserializer.deserialize_str(PeerDidVisitor(PhantomData))
+    }
+}
```

Comment:
The `CowStr` type from `shared_vcx` could be used here for less boilerplate deserialization. If the input can be borrowed from, it will borrow it and `.as_ref()` could be used to get a `&str`. Otherwise it will make an owned variant of the data, which again can be turned into a `&str` through the same method.

Reply:
We already had a discussion about bringing in dependencies unnecessarily, and you know my opinion on it.